File: ordermanager-backend/src/main/java/com/pr/ordermanager/invoice/service/InvoiceMapper.java
package com.pr.ordermanager.invoice.service;

import com.pr.ordermanager.common.model.DropdownDataType;
import com.pr.ordermanager.invoice.entity.Invoice;
import com.pr.ordermanager.invoice.entity.InvoiceItem;
import com.pr.ordermanager.invoice.entity.ItemCatalog;
import com.pr.ordermanager.invoice.entity.RateType;
import com.pr.ordermanager.invoice.model.InvoiceFormModel;
import com.pr.ordermanager.invoice.model.InvoiceItemModel;
import com.pr.ordermanager.invoice.model.ItemCatalogModel;
import com.pr.ordermanager.utils.Utils;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingConstants;
import org.mapstruct.MappingTarget;
import org.mapstruct.ReportingPolicy;

import java.util.List;
import java.util.stream.Collectors;

@Mapper(componentModel = MappingConstants.ComponentModel.SPRING, unmappedTargetPolicy = ReportingPolicy.IGNORE)
public interface InvoiceMapper {

    @Mapping(target = "rateType", expression = "java(RateType.valueOf(invoiceFormModel.getRateType()))")
    @Mapping(target = "invoiceSupplierPerson", ignore = true)
    @Mapping(target = "invoiceRecipientPerson", ignore = true)
    @Mapping(target = "invoiceItems", ignore = true)
    @Mapping(target = "invoiceUser", ignore = true)
    Invoice mapInvoiceModelToEntity(InvoiceFormModel invoiceFormModel);

    @Mapping(target = "rateType", expression = "java(RateType.valueOf(invoiceFormModel.getRateType()))")
    @Mapping(target = "invoiceSupplierPerson", ignore = true)
    @Mapping(target = "invoiceRecipientPerson", ignore = true)
    @Mapping(target = "invoiceItems", ignore = true)
    @Mapping(target = "invoiceUser", ignore = true)
    void mapInvoiceModelToExistedEntity(InvoiceFormModel invoiceFormModel, @MappingTarget Invoice invoice);

    @Mapping(target = "itemCatalog", ignore = true)
    @Mapping(target = "invoice", ignore = true)
    InvoiceItem mapInvoiceItemModelToEntity(InvoiceItemModel invoiceItemModel);

    ItemCatalog mapModelToItemCatalogEntity(ItemCatalogModel source);

    void mapItemCatalogModelToExistedEntity(ItemCatalogModel model, @MappingTarget ItemCatalog entity);

    @Mapping(target = "personRecipientId", source = "invoiceRecipientPerson.id")
    @Mapping(target = "personSupplierId", source = "invoiceSupplierPerson.id")
    @Mapping(target = "rateType", expression = "java(source.getRateType().name())")
    @Mapping(target = "recipientFullName", expression = "java(fullName(source.getInvoiceRecipientPerson().getPersonFirstName(), source.getInvoiceRecipientPerson().getPersonLastName(), source.getInvoiceRecipientPerson().getCompanyName()))")
    @Mapping(target = "supplierFullName", expression = "java(fullName(source.getInvoiceSupplierPerson().getPersonFirstName(), source.getInvoiceSupplierPerson().getPersonLastName(), source.getInvoiceSupplierPerson().getCompanyName()))")
    @Mapping(target = "invoiceItems", expression = "java(source.getInvoiceItems().stream().map(this::mapEntityToModelInvoiceItem).collect(Collectors.toList()))")
    InvoiceFormModel mapInvoiceEntityToFormModel(Invoice source);

    @Mapping(target = "catalogItemId", source = "itemCatalog.id")
    @Mapping(target = "description", source = "itemCatalog.description")
    InvoiceItemModel mapEntityToModelInvoiceItem(InvoiceItem source);

    ItemCatalogModel mapEntityToItemCatalogModel(ItemCatalog itemCatalog);

    default List<DropdownDataType> mapListCatalogItemsToDropdownType(List<ItemCatalog> itemCatalogs) {
        return itemCatalogs.stream().map(c -> new DropdownDataType(
                c.getDescription() + " : Price :" + c.getItemPrice() + " ",
                String.valueOf(c.getId()))).collect(Collectors.toList());
    }

    default String fullName(String firstName, String lastName, String companyName) {
        return (Utils.emptyOrValue(firstName) + " " + Utils.emptyOrValue(lastName) + " " + Utils.emptyOrValue(companyName)).trim();
    }
}

File: ordermanager-backend/src-node/src/invoice/mappers/invoice.mapper.ts
import { Injectable } from '@nestjs/common';
import { DropdownDataTypeDto } from '../../common/dto/common.dto';
import { InvoiceEntity } from '../entities/invoice.entity';
import { InvoiceItemEntity } from '../entities/invoice-item.entity';
import { ItemCatalogEntity } from '../entities/item-catalog.entity';
import { InvoiceFormModelDto, InvoiceItemModelDto, ItemCatalogModelDto } from '../dto/invoice.dto';

const empty = (v?: string | null) => v ?? '';

@Injectable()
export class InvoiceMapper {
  mapInvoiceEntityToFormModel(source: InvoiceEntity): InvoiceFormModelDto {

    const dto: InvoiceFormModelDto = new InvoiceFormModelDto();
    dto.id = source.id;
    dto.invoiceNumber = source.invoiceNumber;
    dto.invoiceDescription = source.invoiceDescription;
    dto.personSupplierId = source.invoiceRecipientPerson?.id??null
    dto.personSupplierId = source.invoiceRecipientPerson?.id??null
    dto.recipientFullName = empty(source?.invoiceRecipientPerson?.personFirstName??'').concat(empty(source.invoiceRecipientPerson?.personLastName??'')).concat(empty(source.invoiceRecipientPerson?.companyName??'')).trim()
    dto.supplierFullName = empty(source?.invoiceSupplierPerson?.personFirstName??'').concat(empty(source?.invoiceSupplierPerson?.personLastName??'')).concat(empty(source.invoiceSupplierPerson?.companyName??'')).trim()
    dto.totalSumNetto = source.totalSumNetto
    dto.totalSumNetto = source.totalSumNetto
    dto.totalSumBrutto = source.totalSumBrutto
    dto.creationDate = new Date(source.creationDate).toISOString()
    dto.invoiceDate = new Date(source.invoiceDate).toISOString()
    dto.rateType = source.rateType
    dto.invoiceItems = (source.invoiceItems ?? []).map((it) => this.mapEntityToModelInvoiceItem(it))

    return dto;
  }

  mapEntityToModelInvoiceItem(source: InvoiceItemEntity): InvoiceItemModelDto {
    const dto: InvoiceItemModelDto = new InvoiceItemModelDto()
    dto.id = source.id
    dto.catalogItemId = source.itemCatalog?.id
    dto.description = source.itemCatalog?.description
    dto.amountItems = source.amountItems
    dto.itemPrice = source.itemPrice
    dto.vat = source.vat
    dto.sumNetto = source.sumNetto
    dto.sumBrutto = source.sumBrutto

    return dto;
  }

  mapEntityToItemCatalogModel(itemCatalog: ItemCatalogEntity): ItemCatalogModelDto {
    return {
      id: itemCatalog.id,
      description: itemCatalog.description,
      shortDescription: itemCatalog.shortDescription,
      itemPrice: itemCatalog.itemPrice,
      vat: itemCatalog.vat,
    };
  }

  mapListCatalogItemsToDropdownType(itemCatalogs: ItemCatalogEntity[]): DropdownDataTypeDto[] {
    return itemCatalogs.map((c) => new DropdownDataTypeDto(`${c.description} : Price :${c.itemPrice} `, String(c.id)));
  }
}
